File: dd-web-full-test/scripts/preflight.py
#!/usr/bin/env python3
"""Pre-flight check for dd-web-full-test. Outputs JSON."""
import json, os, subprocess

PKG = "package.json"
result = {
    "framework": "unknown", "buildTool": "unknown",
    "devServerPort": "3000", "devServerRunning": False,
    "tools": {"playwright": None, "vitest": None, "jest": None, "axe": False},
    "componentLib": None, "missingTools": [],
    "runner": "python"
}

# ── Detect framework & build tool ──────────────────────────────────

if os.path.exists(PKG):
    try:
        pkg = json.load(open(PKG)); deps = {**pkg.get("dependencies",{}), **pkg.get("devDependencies",{})}
        if "next" in deps: result["framework"], result["buildTool"], result["componentLib"] = "next.js", "next.js", "@testing-library/react"
        elif "react" in deps: result["framework"], result["componentLib"] = "react", "@testing-library/react"
        elif "vue" in deps: result["framework"], result["componentLib"] = ("vue3" if str(deps["vue"]).startswith(("3","^3")) else "vue2"), "@testing-library/vue"
        elif "@angular/core" in deps: result["framework"], result["componentLib"] = "angular", "@angular/core/testing"
        elif "svelte" in deps: result["framework"], result["componentLib"] = "svelte", "@testing-library/svelte"
        elif "solid-js" in deps: result["framework"], result["componentLib"] = "solidjs", "@solidjs/testing-library"
        if os.path.exists("vite.config.js") or os.path.exists("vite.config.ts"): result["buildTool"] = "vite"
        elif os.path.exists("angular.json"): result["buildTool"] = "angular-cli"
        elif os.path.exists("svelte.config.js"): result["buildTool"] = "sveltekit"
        elif "react-scripts" in deps: result["buildTool"] = "create-react-app"
        scripts = pkg.get("scripts",{}); dc = scripts.get("dev","") or scripts.get("start","")
        import re; m = re.search(r'(?:--port|-p)\s+(\d+)', dc)
        if m: result["devServerPort"] = m.group(1)
    except: pass
else:
    result["framework"], result["buildTool"] = "vanilla", "none"

# ── Check tools ────────────────────────────────────────────────────

def ver(*commands):
    for cmd in commands:
        try:
            out = subprocess.check_output(cmd, stderr=subprocess.DEVNULL, timeout=5)
            return out.decode().strip().split("\n")[-1]
        except Exception:
            pass
    return None

result["tools"]["playwright"] = ver(["node_modules/.bin/playwright", "--version"], ["playwright", "--version"])
result["tools"]["vitest"] = ver(["node_modules/.bin/vitest", "--version"], ["vitest", "--version"])
result["tools"]["jest"] = ver(["node_modules/.bin/jest", "--version"], ["jest", "--version"])
result["tools"]["axe"] = os.path.exists("node_modules/@axe-core/playwright")

try:
    import urllib.request; urllib.request.urlopen(f"http://localhost:{result['devServerPort']}", timeout=2); result["devServerRunning"] = True
except: pass

if not result["tools"]["playwright"]: result["missingTools"].append("@playwright/test")
if not result["tools"]["vitest"] and not result["tools"]["jest"]: result["missingTools"].append("vitest")
if not result["tools"]["axe"]: result["missingTools"].append("@axe-core/playwright")

print(json.dumps(result, indent=2))

File: dd-web-full-test/scripts/preflight.mjs
#!/usr/bin/env node
/**
 * Pre-flight check for dd-web-full-test.
 * Detects framework, build tool, installed tools. Outputs JSON.
 * Usage: node scripts/preflight.mjs
 */
import { readFileSync, existsSync } from "node:fs";
import { execSync } from "node:child_process";
import { request } from "node:http";

const result = {
  framework: "vanilla",
  buildTool: "none",
  devServerPort: "3000",
  devServerRunning: false,
  tools: { playwright: null, vitest: null, jest: null, axe: false },
  componentLib: null,
  missingTools: [],
  runner: "node",
};

// ── Detect framework & build tool ──────────────────────────────────

if (existsSync("package.json")) {
  try {
    const pkg = JSON.parse(readFileSync("package.json", "utf-8"));
    const deps = { ...pkg.dependencies, ...pkg.devDependencies };

    if (deps.next) { result.framework = "next.js"; result.buildTool = "next.js"; result.componentLib = "@testing-library/react"; }
    else if (deps.react) { result.framework = "react"; result.componentLib = "@testing-library/react"; }
    else if (deps.vue) { result.framework = String(deps.vue).startsWith("2") ? "vue2" : "vue3"; result.componentLib = "@testing-library/vue"; }
    else if (deps["@angular/core"]) { result.framework = "angular"; result.componentLib = "@angular/core/testing"; }
    else if (deps.svelte) { result.framework = "svelte"; result.componentLib = "@testing-library/svelte"; }
    else if (deps["solid-js"]) { result.framework = "solidjs"; result.componentLib = "@solidjs/testing-library"; }

    if (existsSync("vite.config.js") || existsSync("vite.config.ts")) result.buildTool = "vite";
    else if (existsSync("angular.json")) result.buildTool = "angular-cli";
    else if (existsSync("svelte.config.js")) result.buildTool = "sveltekit";
    else if (deps["react-scripts"]) result.buildTool = "create-react-app";

    const scripts = pkg.scripts || {};
    const devCmd = scripts.dev || scripts.start || "";
    const portMatch = devCmd.match(/(?:--port|-p)\s+(\d+)/);
    if (portMatch) result.devServerPort = portMatch[1];
  } catch { /* ignore parse errors */ }
}

// ── Check tools ────────────────────────────────────────────────────

function getVersion(cmd) {
  try {
    return execSync(cmd, { timeout: 5000, stdio: ["ignore", "pipe", "ignore"] })
      .toString().trim().split("\n").pop();
  } catch { return null; }
}

result.tools.playwright = getVersion("node_modules/.bin/playwright --version") || getVersion("playwright --version");
result.tools.vitest = getVersion("node_modules/.bin/vitest --version") || getVersion("vitest --version");
result.tools.jest = getVersion("node_modules/.bin/jest --version") || getVersion("jest --version");
result.tools.axe = existsSync("node_modules/@axe-core/playwright");

// Dev server check — sync, blocks max 2s. Sets devServerRunning to true if reachable.
try {
  await new Promise((resolve) => {
    const req = request(
      { hostname: "localhost", port: parseInt(result.devServerPort), path: "/", method: "HEAD", timeout: 2000 },
      (res) => { result.devServerRunning = true; res.resume(); resolve(); }
    );
    req.on("error", () => resolve());
    req.on("timeout", () => { req.destroy(); resolve(); });
    req.end();
    setTimeout(() => { req.destroy(); resolve(); }, 2000);
  });
} catch { /* offline */ }

// Missing
if (!result.tools.playwright) result.missingTools.push("@playwright/test");
if (!result.tools.vitest && !result.tools.jest) result.missingTools.push("vitest");
if (!result.tools.axe) result.missingTools.push("@axe-core/playwright");

console.log(JSON.stringify(result, null, 2));
